### DISCUSSION TITLE
Change Filament Fails In Compiled Manual Mesh 3X3

### DISCUSSION THREAD

**Discussion by Cal-LF** [Jyers/Marlin#597] [Q&A] [answered]
When using the precompiled manual mesh 3X3, change filament works as it should. When compiling the manual mesh 3X3 template, filament change does an extra unload of filament of about 40mm as it is returning to the print. It then continues as normal but no filament will come out because it has backed out to the top of the hot end coupler.
My previous compiles of this template were working for filament change.

**Comment by Jyers** [comment]
Nothing has changed regarding those settings in the config. On top of this, the precompiled versions are built from the template configs. My best guess would be an eeprom issue, as I have seen that affect filament changes before. Try again after resetting and storing to eeprom

**Reply by Cal-LF** [reply]
Ok, thanks. I will try reset again after a new compile.

**Reply by Cal-LF** [reply]
Self compiled firmware still does not run the filament change or filament runout without retracting filament right after coming off park and back to the print. The precompiled versions work as they should. The eprom reset has no effect. I tried 2 different templates to compile. Tried 2 different precompiles and they work properly.
At this point I will use precompiled. Will play with settings in the configuration files in the compiler to see what I am doing wrong. The most simple compile of changing only the motherboard to V427 still has failed filament change and filament runout.

**Reply by Jyers** [reply]
I checked my github workflows for sanities sake and yup, the precompiled manual mesh versions are built from the manual mesh config templates. I am baffled by what might be causing this...

**Reply by Cal-LF** [reply]
Thanks Jake. I will eventually find what is going on. Your personal responses are greatly appreciated. I like messing around with firmware. If there were no issued to figure out, it would be boring.
This discussion can be closed.

**Comment by Cal-LF** [comment, marked as answer]
The work-around is to push the filament back to the nozzle before it starts printing again. I am there anyways changing the filament so just have to remember to do it quickly. The steppers are locked so moving the X and Z axis is not likely.
Also found that Pause Print is faster to change filament.

**Comment by xantrk** [comment]
@Jyers Im not sure if it is related but my filament change does some weid extra purge(?) after I suppose everything is finished. But it tries to purge very fast and many extruder skips for ~10 seconds. I checked the config and significantly lowered feedrates in config but it still does that but not always. Could it be an issue with extruder staying in relative mode, but m600 uses absolute? 

I checked the code, I believe that menu is only sending m600 command , nothing else. So could there be an issue with Marlin itself?

**Reply by Cal-LF** [reply]
The load length needs to be reduced until it stops trying to fast feed. Mine did the same thing.

**Reply by xantrk** [reply]
> The load length needs to be reduced until it stops trying to fast feed. Mine did the same thing.

It was 0. But maybe it has an internal default. I will try a couple of mm load only. But it appeared after a considerable amount of purge, some pause then it went crazy.  I was thinking it was after; unload, fast load, slow load, purge, pause for couple of seconds then crazy stuff but not sure.

**Reply by xantrk** [reply]
@Cal-LF May I ask if you are using prusaslicer (default relative extrusion), or cura (default absolute extrusion)? 

@Jyers I really think there is something wrong with m600. Last two times I tried to change filament after a powercycle, everything works fine. But if I do that after a print; it unloads, slow loads, fast loads then purges fine, waits around 10 seconds then extruder skips and purges like crazy, while at the same time y axis slowly moves. I had similar skipping when I tried to switch to relative extrusion in prusaslicer, but forgot to delete my layer change gcode so every layer it purged and skipped.

I also tried to set slow and fast load distances to 10mm and feedrate to 3, reflashed and resetted eeprom. Everything is easy and slow, but It is definitely doing something after a very slow purge and 10 second wait, it tries to move really fast and skips a lot. Any ideas on what should I do or check?

**Reply by xantrk** [reply]
https://forum.makerforums.info/t/solved-gcode-for-filament-swap-m600/77922 this looks relevant

**Reply by xantrk** [reply]
> Also try to calibrate your extruder correctly! An extruder that skips is often an improperly calibrated extruder. By default the Creality settings are not good.
> A very good site on this subject: https://teachingtechyt.github.io/calibration.html

I recently have done it already. 

The issue I mentioned was another relative/absolute extruder misplacement issue I had, where print is using absolute steps, but layer change code resets the extruder as if I am using relative. So that for example first layer finishes at E17.0, change gode resets the extruder, then next gcode command extrudes to 17.1 but from 0 (because resetted) rather than 17.0 and only extruding 0.1 mm. Hope I explained clearly. 

The issue is like somehow printed tries to restore an initial state after filament change is done, but the extruder is far off. I am saying that because the issue only appears using the iflament change after a print, an is fine after powercycling.

**Reply by Jyers** [reply]
@xantrk Could be an issue with the gcode, if it's an issue with the firmware than you'll want to report it on the main marlin repo, not here.

**Reply by xantrk** [reply]
> @xantrk Could be an issue with the gcode, if it's an issue with the firmware than you'll want to report it on the main marlin repo, not here.

It turns out this is a marlin issue, which was fixed with https://github.com/MarlinFirmware/Marlin/pull/21670/commits/1efe8a546dc9739597a000c2729e6a379faccd66 . I believe it will be fixed when upstream branch is pulled. Hope we didn't bother you a lot :)

**Reply by Ka8oum** [reply]
I am not a github pro so I don't really know how to pull the upstream branch myself, but indeed I copied the change into my jyers's local copy in the pause.cpp file, compiled, and it works as expected now

**Comment by xantrk** [comment]
Thanks Jyers, will do that. I realized that the function  that is called to
apply gcode, tries to restore planning state. Checked the gcode, pure m600
works fine. If you think this is unrelated, will report to Marlin repo.
